src/box/box.cc

```diff
@@ -4448,7 +4469,7 @@ box_process_fetch_snapshot(struct iostream *io,
 	/* Send the snapshot data to the instance. */
 	struct vclock start_vclock;
 	relay_initial_join(io, header->sync, &start_vclock, req.version_id,
-			   cursor_ptr);
+			   cursor_ptr, replica);
```

Comment:
I don't understand why we need to pass `replica` to `relay_initial_join`. We don't do it in `box_process_join` although the replica may exist there as well.

Reply:
Hm, actually we should do it there as well - it's my mistake. The point is we determine if a replica is connected by the state of its relay so we should replica's relay as often as possible so that another connection with the same UUID will be rejected.

After your comment I actually realised that this protection is like a sieve. If another replica will try to join with the same UUID when relay is not used (on `_cluster:insert`, for example), master will fail. Probably, we should do a better protection against multiple connections? For example, something like `RegionGuard` but for replicas that will be used in all four replication endpoints.

Maybe the problem with join is not so critical since only regular replicas use it so the protocol is kind of "closed". But what's about anonymous replicas, anyone can write its own and use the protocol in his way. For example, one can send a new fetch_snapshot request with the same UUID right after the previous one failed because of the network failure. However, Tarantool may notice the disconnection a bit later so it's actually a duplicate connection from Tarantool's point of view - and not rejected duplicate connection is likely to crash the Tarantool.

Reply:
1. Removed the replica from `relay_initial_join` (reverted my change, actually).
2. Introduced `ReplicaConnectionGuard` for better protection against duplicate connections.